src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm

```diff
@@ -15,325 +15,285 @@
+{
+    NSParameterAssert(duplicateCheckHandler);
+    [self assertMutable];
+    _duplicateTypeID = opaqueDuplicateTypeID;
+    _duplicateCheckHandler = duplicateCheckHandler;
+}
+
 @end
 
-@implementation MTRAsyncWorkItem
+MTR_DIRECT_MEMBERS
+@implementation MTRAsyncWorkQueue {
+    os_unfair_lock _lock;
+    __weak id _context;
+    NSMutableArray<MTRAsyncWorkItem *> * _items;
+    NSInteger _runningWorkItemCount;
+}
 
-- (instancetype)initWithQueue:(dispatch_queue_t)queue
+- (instancetype)initWithContext:(id)context
 {
+    NSParameterAssert(context);
     if (self = [super init]) {
-        _lock = OS_UNFAIR_LOCK_INIT;
-        _queue = queue;
+        _context = context;
+        _items = [NSMutableArray array];
     }
     return self;
 }
 
-// assume lock is held
-- (void)_invalidate
-{
-    // Make sure we don't leak via handlers that close over us, as ours must.
-    // This is a bit odd, since these are supposed to be non-nullable
-    // properties, but it's the best we can do given our API surface, unless we
-    // assume that all consumers consistently use __weak refs to us inside their
-    // handlers.
-    //
-    // Setting the attributes to nil will not compile; set the ivars directly.
-    _readyHandler = nil;
-    _cancelHandler = nil;
-}
-
-- (void)invalidate
+- (NSString *)description
 {
     os_unfair_lock_lock(&_lock);
-    [self _invalidate];
+    auto * result = [NSString stringWithFormat:@"<%@ context: %@ items count: %tu>", self.class, _context, _items.count];
     os_unfair_lock_unlock(&_lock);
+    return result;
 }
 
-- (void)markedEnqueued
+- (void)enqueueWorkItem:(MTRAsyncWorkItem *)item
 {
+    NSParameterAssert(item);
+    NSAssert(_context, @"context has been lost");
+    [item markedEnqueued];
+
     os_unfair_lock_lock(&_lock);
-    _enqueued = YES;
+    [_items addObject:item];
+    [self _callNextReadyWorkItem];
     os_unfair_lock_unlock(&_lock);
 }
 
-- (void)setReadyHandler:(MTRAsyncWorkReadyHandler)readyHandler
+- (void)invalidate
 {
     os_unfair_lock_lock(&_lock);
-    if (!_enqueued) {
-        _readyHandler = readyHandler;
+    MTR_LOG_INFO("MTRAsyncWorkQueue<%@> invalidate %tu items", _context, _items.count);
+    for (MTRAsyncWorkItem * item in _items) {
+        [item cancel];
     }
+    [_items removeAllObjects];
     os_unfair_lock_unlock(&_lock);
 }
 
-- (void)setCancelHandler:(dispatch_block_t)cancelHandler
+- (void)_postProcessWorkItem:(MTRAsyncWorkItem *)workItem retry:(BOOL)retry
 {
-    os_unfair_lock_lock(&_lock);
-    if (!_enqueued) {
-        _cancelHandler = cancelHandler;
+    MTRAsyncWorkItem * runningWorkItem = (_runningWorkItemCount) ? _items.firstObject : nil;
+    if (workItem != runningWorkItem) {
+        NSAssert(NO, @"work item to post-process is not running");
+        return;
     }
-    os_unfair_lock_unlock(&_lock);
-}
 
-- (void)endWork
-{
-    [self.workQueue endWork:self];
-    [self invalidate];
```

Comment:
This is hard to parse because the whole file got reordered so the diff is not very readable, but I believe this used to drop references to the handlers, and we are no longer doing that.  

The dropping references was a cycle-breaking measure, but it still seems like a good idea, no?  As things stand we could have handlers that accidentally close over the work item and then we will leak.

Concretely, at the points when we mark a work item complete we should drop all its refs to handlers.

Reply:
I can add that back in... I really dislike this "generic object with a bunch of blocks" instead of "sub-classing and overriding the relevant methods" pattern, because it produces all these sorts of issues. If only ObjC had anonymous sub-classes...